File: samples/java/SupportChat/Server/Support/src/main/java/systems/zlink/samples/supportchat/server/support/spots/entryspot/SupportEntrySpot.java
package systems.zlink.samples.supportchat.server.support.spots.entryspot;

import systems.zlink.framework.messaging.ZLinkMessage;
import systems.zlink.framework.spots.ZLinkActorCreateResponse;
import systems.zlink.framework.spots.ZLinkEntrySpot;
import systems.zlink.framework.spots.ZLinkEntrySpotContext;
import systems.zlink.samples.supportchat.server.configuration.SampleNames;
import systems.zlink.samples.supportchat.server.support.actors.SupportActorDirectory;
import systems.zlink.samples.supportchat.server.support.actors.SupportUserActor;
import systems.zlink.samples.supportchat.server.support.application.AgentAssignmentService;
import systems.zlink.samples.supportchat.shared.contracts.Messages;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;

public final class SupportEntrySpot implements ZLinkEntrySpot<SupportUserActor> {
    private final ZLinkEntrySpotContext context;
    private final SupportActorDirectory directory;
    private final AgentAssignmentService assignment;

    public SupportEntrySpot(
            ZLinkEntrySpotContext context,
            SupportActorDirectory directory,
            AgentAssignmentService assignment) {
        this.context = context;
        this.directory = directory;
        this.assignment = assignment;
    }

    @Override
    public ZLinkEntrySpotContext context() {
        return context;
    }

    @Override
    public CompletionStage<ZLinkActorCreateResponse> onCreateActor(
            SupportUserActor actor, ZLinkMessage createRequest) {
        Messages.EnsureSupportUserActorReq request =
                createRequest.decode(Messages.EnsureSupportUserActorReq.class);
        actor.setIdentity(request.displayName(), request.role(), request.participantId());
        directory.remember(actor);
        return CompletableFuture.completedFuture(ZLinkActorCreateResponse.accept());
    }

    @Override
    public CompletionStage<Void> onJoinedActor(SupportUserActor actor) {
        return CompletableFuture.completedFuture(null);
    }

    @Override
    public CompletionStage<Void> onLeaveActor(SupportUserActor actor) {
        return CompletableFuture.completedFuture(null);
    }

    // --8<-- [start:doc-sc-agent-disconnect]
    @Override
    public CompletionStage<Void> onDisconnectActor(SupportUserActor actor) {
        if (SampleNames.Roles.Agent.equals(actor.role())) {
            assignment.setAvailable(actor.participantId(), actor.displayName(), false);
        }
        return CompletableFuture.completedFuture(null);
    }
    // --8<-- [end:doc-sc-agent-disconnect]
}

File: samples/kotlin/SupportChat/Server/Support/src/main/kotlin/systems/zlink/samples/kotlin/supportchat/server/support/infrastructure/zlink/spots/entryspot/SupportEntrySpot.kt
package systems.zlink.samples.kotlin.supportchat.server.support.infrastructure.zlink.spots.entryspot

import systems.zlink.framework.actors.ZLinkActorManager
import systems.zlink.framework.kotlin.ZLinkSuspendingEntrySpot
import systems.zlink.framework.kotlin.await
import systems.zlink.framework.messaging.ZLinkMessage
import systems.zlink.framework.spots.ZLinkActorCreateResponse
import systems.zlink.framework.spots.ZLinkEntrySpotContext
import systems.zlink.samples.kotlin.supportchat.server.configuration.SupportChatRoles
import systems.zlink.samples.kotlin.supportchat.server.support.application.AgentAssignmentService
import systems.zlink.samples.kotlin.supportchat.server.support.infrastructure.zlink.actors.SupportActorDirectory
import systems.zlink.samples.kotlin.supportchat.server.support.infrastructure.zlink.actors.SupportUserActor
import systems.zlink.samples.kotlin.supportchat.shared.contracts.EnsureSupportUserActorReq

class SupportEntrySpot(
    override val context: ZLinkEntrySpotContext,
    private val directory: SupportActorDirectory,
    private val assignment: AgentAssignmentService,
    private val actorManager: ZLinkActorManager,
) : ZLinkSuspendingEntrySpot<SupportUserActor>() {
    override suspend fun onCreateActorSuspending(
        actor: SupportUserActor,
        createRequest: ZLinkMessage,
    ): ZLinkActorCreateResponse {
        val request = createRequest.decode(EnsureSupportUserActorReq::class.java)
        actor.setIdentity(request.displayName, request.role, request.participantId)
        val actorRef =
            actorManager.find(actor.actorId).await().orElse(null)
                ?: throw IllegalStateException(
                    "Support actor ref is not available. actor=${actor.actorId}"
                )
        directory.addOrUpdate(actor, actorRef)
        return ZLinkActorCreateResponse.accept()
    }

    override suspend fun onJoinedActorSuspending(actor: SupportUserActor) {}

    override suspend fun onLeaveActorSuspending(actor: SupportUserActor) {}

    // --8<-- [start:doc-sc-agent-disconnect]
    override suspend fun onDisconnectActorSuspending(actor: SupportUserActor) {
        if (actor.role == SupportChatRoles.Agent) {
            assignment.setAvailable(actor.actorId, actor.displayName, false)
        }
    }
    // --8<-- [end:doc-sc-agent-disconnect]
}
